javaslang/generator/Generator.scala

```diff
@@ -51,6 +51,377 @@ def generateMainClasses(): Unit = {
+           */
+          public static <T> $FutureType<T> future(T result) {
+            return $FutureType.successful(result);
+          }
+
+          /$javadoc
+           * Alias for {@link $FutureType#successful($ExecutorServiceType, Object)}
+           *
+           * @param <T>             The value type of a successful result.
+           * @param executorService An {@code ExecutorService}.
+           * @param result          The result.
+           * @return A succeeded {@code Future}.
+           * @throws NullPointerException if executorService is null
+           */
+          public static <T> $FutureType<T> future($ExecutorServiceType executorService, T result) {
+            return $FutureType.successful(executorService, result);
+          }
+
+          /$javadoc
+           * Alias for {@link $FutureType#failed(Throwable)}
+           *
+           * @param <T>       The value type of a successful result.
+           * @param exception The reason why it failed.
+           * @return A failed {@code Future}.
+           * @throws NullPointerException if exception is null
+           */
+          public static <T> $FutureType<T> future(Throwable exception) {
+            return $FutureType.failed(exception);
+          }
+
+          /$javadoc
+           * Alias for {@link $FutureType#failed($ExecutorServiceType, Throwable)}
+           *
+           * @param <T>             The value type of a successful result.
+           * @param executorService An executor service.
+           * @param exception       The reason why it failed.
+           * @return A failed {@code Future}.
+           * @throws NullPointerException if executorService or exception is null
+           */
+          public static <T> $FutureType<T> future($ExecutorServiceType executorService, Throwable exception) {
+            return $FutureType.failed(executorService, exception);
+          }
+
+          /$javadoc
+           * Alias for {@link Lazy#of($SupplierType)}
+           *
+           * @param <T>      type of the lazy value
+           * @param supplier A supplier
+           * @return A new instance of Lazy
+           */
+          public static <T> Lazy<T> lazy($SupplierType<? extends T> supplier) {
+            return Lazy.of(supplier);
+          }
+
+          /$javadoc
+           * Alias for {@link Option#of(Object)}
+           *
+           * @param <T>   type of the value
+           * @param value A value
+           * @return {@code Some(value)} if value is not {@code null}, {@code None} otherwise
+           */
+          public static <T> $OptionType<T> option(T value) {
+            return $OptionType.of(value);
+          }
+
+          /$javadoc
+           * Alias for {@link Option#some(Object)}
+           *
+           * @param <T>   type of the value
+           * @param value A value
+           * @return {@code Some(value)}
+           */
+          public static <T> $OptionType<T> some(T value) {
```

Comment:
This is a valid point. Several types already have the static factory methods with similar names:

* [x] controls (Either, Option, Try, Validation)
* [x] Future has successful() and failed() (Promise has the same method names...)
* [ ] Collections
* [ ] Tuples
* [ ] Functions

We still have additional problems here - some API is ambiguous, e.g.

```java
// returns a success
public static <T> $FutureType<T> future(T result) { ... }

// returns a failure
public static <T> $FutureType<T> future(Throwable exception) { ... }
```

It was a design decision to use consequently use `of` throughout our whole codebase, e.g.

* `Tuple.of(1, true, "nice")`
* `List.of(1, 2, 3)`
* `Try.of(() -> computation())`

* Java also uses the `of` API. It looks familiar to the user.
* Lower case 'constructors' look alien to Java developers.
* Scala has companion objects which give use upper case constructors without the use of `new` - that is what we really want. But the upper-case names are already taken by the Pattern Match API in order to deconstruct objects.
* Minor: Static imports are always a few keystrokes more for the developer, even with an IDE.

**But: The `of` notation is not concise enough.**

Like @paplorinc said, we do not need to duplicate the existing API. What we really want are shortcuts for the `of` factory methods only.

I would like to have first char **upper-case** static factory method names within `javaslang.API`, e.g.

* `Tuple(1, 2)`, `Tuple(1, 2, 3, 4)`
* `Function1(obj::methodRef)`, `CheckedFunction1(obj::methodRef)`
* `Try(() -> computation())`, `Option(value)`, `Future(() -> computation())`
* `List(1, 2, 3)`
* `HashMap(1, "a", 2, "b", 3, "c")`

This would go adhere with `API.Match()` and `API.Case()` etc.

To this day static factory methods for (most) collection _interfaces_ were out of scope, e.g.

* `Seq(1, 2, 3)`
* `Set(1, 2, 3)`
* `Map(1, "a", 2, "b", 3, "c")`

But maybe it would be nice to have these too.

---

**Q:** So, what to do with the names of generated patterns? They will clash with the new API.

**A:** We will break backward compatibility and change the annotation processor of `javaslang-match`. We will generate a preceding `$` for patterns. This goes adhere with the use of `$` for standard patterns like

* `$()` - any
* `$(value)` - equality
* `$(predicate)` - condition

Example:

```java
T result = Match(value).of(
    Case($Tuple($(1), $(true), $("nice")), (v1, v2, v3) -> ...),
    Case($List($(1), $()), (head, tail) -> ...)
);
```

We determine these _specialities_:

* `$Tuple($(Tuple(1, 2, 3))` will contain a `$Tuple` pattern and a `Tuple` constructor. Hard to read?
* `$List($(), $())` deconstructs _head_ element and _tail_ list
* `List(v1, v2)` constructs a List containing two values `v1`, `v2`

---

# The bottom line is that this will be a breaking change that will be targeted for 3.0.0

However, the first char upper case constructors can be already introduced but they will raise problems when used in conjunction with the Match API (, e.g. `import static javaslang.API.*;`).